core/dbt/contracts/graph/manifest.py

```diff
@@ -145,19 +145,22 @@ def perform_lookup(self, unique_id: UniqueID, manifest: "Manifest") -> SourceDef
 class RefableLookup(dbtClassMixin):
     # model, seed, snapshot
     _lookup_types: ClassVar[set] = set(NodeType.refable())
+    _versioned_types: ClassVar[set] = set(NodeType.versioned())
 
     # refables are actually unique, so the Dict[PackageName, UniqueID] will
     # only ever have exactly one value, but doing 3 dict lookups instead of 1
     # is not a big deal at all and retains consistency
     def __init__(self, manifest: "Manifest"):
-        self.storage: Dict[str, Dict[PackageName, UniqueID]] = {}
+        self.storage: Dict[str, Dict[PackageName, List[UniqueID]]] = {}
```

Comment:
`ref` lookups now need to handle 3 access patterns (`package` handling remains unchanged):
1. Lookup versioned model by name and version => lookup specified version
2. Lookup versioned model by name _without_ version => get the latest version
3. Lookup unversioned model by name only => get only 'version'

For example, a manifest with the following nodes: 
```json
"model.jaffle_shop.dim_unversioned": {
  "name": "dim_unversioned",
  "version": null, 
  "is_latest_version": null
},
"model.jaffle_shop.dim_customers.v1": {
  "name": "dim_customers",
  "version": "1", 
  "is_latest_version": false
},
"model.jaffle_shop.dim_customers.v2": {
  "name": "dim_customers",
  "version": "2", 
  "is_latest_version": true
}
```
should find unique_ids for the following refs via `RefableLookup.get_unique_id`:
```
ref("dim_unversioned") => "model.jaffle_shop.dim_unversioned"
ref("dim_customers", version="1") => "model.jaffle_shop.dim_customers.v1"
ref("dim_customers") => "model.jaffle_shop.dim_customers.v2"
```

Reply:
I considered a few options for the `storage` structure here: 
* Dict[str, Dict[PackageName, List[UniqueID]]]
   * "latest" unique_id stored at the front of list
   * first element of `UniqueID` list always returned, unless a `version` is specified - in which case iterating through the list is necessary.  
   * implemented in this draft
```python
{
"dim_unversioned": {
    "jaffle_shop": ["model.jaffle_shop.dim_unversioned"]
  }, "dim_customers": {
    "jaffle_shop": ["model.jaffle_shop.dim_customers.v2", "model.jaffle_shop.dim_customers.v1"]
  }
}
```
* Dict[str, Dict[PackageName, Dict[VersionName, UniqueID]]]
  * doesn't handle unversioned or "latest" versioned lookups models nicely
```python
{ 
"dim_unversioned": {
  "jaffle_shop": {
    "??": "model.jaffle_shop.dim_unversioned"
  },
},
"dim_customers": {
  "jaffle_shop": {
      "2": "model.jaffle_shop.dim_customers.v2", 
      "1": "model.jaffle_shop.dim_customers.v1"
    }
  }
}
```
* Dict[str, Dict[PackageName, UniqueID]]
  * doesn't handle "latest" versioned lookups models nicely

```python
{ 
"dim_unversioned": {
  "jaffle_shop": model.jaffle_shop.dim_unversioned"
},
"dim_customers.v1": {
  "jaffle_shop": "model.jaffle_shop.dim_customers.v1"
},
"dim_customers.v2": {
  "jaffle_shop": "model.jaffle_shop.dim_customers.v2"
}
}
```

Reply:
For the third option, we _could_ store an additional key here that maps the model name to its latest unique id. For example, 
```

```python
{ 
"dim_unversioned": {
  "jaffle_shop": model.jaffle_shop.dim_unversioned"
},
"dim_customers.v1": {
  "jaffle_shop": "model.jaffle_shop.dim_customers.v1"
},
"dim_customers.v2": {
  "jaffle_shop": "model.jaffle_shop.dim_customers.v2"
},
"dim_customers.
}
```
```
This has the benefits of:
* keeping the storage data structure consistent across different lookup implementations
* mapping closely with the access patterns => easier to reason about + constant-time lookup for all access patterns

Reply:
I assume that the ref_lookup code is going to have to handle external references eventually too, right? Not yet sure how that would impact these design choices.

I am a bit wary of having "latest" only implied by the order of versions in the first option.  That feels a bit fragile to me.  It doesn't feel like checking the versions would be that much overhead. And it's always going to be numerical order, so the highest integer is always latest, right?

Reply:
> I am a bit wary of having "latest" only implied by the order of versions in the first option. That feels a bit fragile to me. 

Same here. 

> And it's always going to be numerical order, so the highest integer is always latest, right?

This is something I haven't put in much polish on for the spike yet but - but we'll be accepting strings for the version identifier, as well as a model-level `latest_version`. If the latest_version is provided - then the latest is that value, regardless of what the highest version is (to support the concept of prereleases). 

If latest_version is not provided, dbt will compute a default latest version by attempting to order the version identifiers numerically if possible (casting to int/float), and fallback to alphanumeric ordering. This is to support a variety of versioning conventions out of the box, even if our recommendation is to use major versioning only. If

Reply:
Yes, I think the latest proposed structure is appealing.